Changed KeychainModule to use Android’s KeyStore for creating and sto…

pcoltau opened the pull request:
…ring encryption keys. Also added functionality to automatically migrate existing encrypted data from Facebook’s conceal to encryption using KeyStore keys.

Signed-off-by: Pelle Stenild Coltau <pcoltau@gmail.com>

MacKentoch commented:
👏 I love this PR!

vonovak commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -45,91 +97,245 @@ public KeychainModule(ReactApplicationContext reactContext) {
+        }
+    }
+
+    private void resetGenericPasswordForOptionsWithDelimiter(String service, String delimiter, boolean resetKeyStore) throws KeyStoreException, KeyStoreAccessException {
```
@pcoltau do we need the boolean parameter? the keystore is reset iff the delimiter is the not-legacy one, right?

pcoltau replied:
Correct. Should I remove the `resetKeyStore` boolean and make a `if (!delimiter.equals(LEGACY_DELIMITER)) {` instead?

vonovak replied:
@pcoltau maybe the best would be to have `resetGenericPasswordForOptions` and `resetGenericPasswordForOptionsLegacy` that both accept only one parameter? Then in future versions when conceal is removed it would be easier to refactor the code..

pcoltau commented:
@vonovak I've updated the code as per your suggestions. Please review again.

vonovak commented:
@pcoltau so the data is still stored in sharedPreferences, the difference being that the private key is somehow stored in the device's hardware? Just so we know the difference. The code looks good to me, I'll test it out when I can.

pcoltau commented:
FYI, I pushed a bug fix, where legacy values with NULL service would not be migrated.

@vonovak Yes. Data is still stored in SharedPreferences, but using a new delimiter (in order to know the difference between new values and legacy values). Encryption is upgraded from Facebook's Conceal that uses 256 bit keys to RSA public/private key encryption using 2048 bit keys. Furthermore the encryption keys are stored in Android's KeyStore. On newer devices, the KeyStore is backed by a hardware chip, so that all keys are generated and stored in hardware. A new key is generated for each unique value of `service`.

pcoltau commented:
I have found an issue with this PR. If the value to be encrypted (username/password) is more than roughly 256 bytes, the encryption will fail. Probably due to the fact that RSA is not a block cipher. I will look into changing it to a block cipher instead (probably AES-CBC).

pcoltau commented:
The RSA implementation has now been replaced with an AES CBC (256 bit) encryption. 

I have not been able to find any other issues. 

Please do review and consider merging to master.

oblador commented:
@pcoltau Amazing work! I'll leave the actual review to @vonovak, but will this cause any breakage of old android versions?

pcoltau commented:
@oblador Very good question regarding the older Android versions. The ability to store AES keys in the Keystore was added in API level 23 (a lot of security features was added in that release). The Keystore API it self was added in API level 18.

I just noticed that this project has the minimum android sdk version set (in `build.gradle`) to 16.

I would highly recommend that we change that to 23 for security concerns.

If you decide to stay at API level 16, then I can make the necessary changes to the code in this PR, so that Facebook's conceal is used in APIs less than 23, but I really would not recommend that.

ggozad commented:
Hey I am the author of https://github.com/Crypho/cordova-plugin-secure-storage
We are gradually migrating to react from cordova and naturally I have an interest in this :)
Let me know if you need any help as we have been using this approach (shared preferences backed by keystore) for a while now.

vonovak commented:
@pcoltau thanks for the good work and sorry for taking so much time. I just replaced the old keychain module with [your branch](https://github.com/TeletronicsDotAe/react-native-keychain/blob/feature/change_to_android_keystore) and got `Error: could not decrypt bytes`. I'll try to take a look at why that is happening so the upgrade is smooth to users. Regarding the switch to API 23 (android 6.0) - that seems to be a rather big jump. I think it'd be ok to switch to API 18 (android 4.3) but switching to 6.0 would make the package unusable for most apps. I think we will need to somehow combine the existing approaches. Thoughts @oblador ?

vonovak commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -45,91 +99,257 @@ public KeychainModule(ReactApplicationContext reactContext) {
+    }
+
+    private byte[] getBytesFromPrefs(String service, String prefix) {
+        String key = service + prefix;
```
@pcoltau this var is unused

vonovak commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -45,91 +99,257 @@ public KeychainModule(ReactApplicationContext reactContext) {
+                    setGenericPasswordForOptions(
+                            originalService,
+                            new String(resultSet.decryptedUsername, Charset.forName("UTF-8")),
+                            new String(resultSet.decryptedUsername, Charset.forName("UTF-8")));
```
@pcoltau this should be decryptedPassword

vonovak commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -45,91 +99,257 @@ public KeychainModule(ReactApplicationContext reactContext) {
+                            new String(resultSet.decryptedUsername, Charset.forName("UTF-8")));
+                    // Remove the legacy value(s)
+                    resetGenericPasswordForOptionsLegacy(originalService);
+                    recuser = resultSet.decryptedUsername;
```
this should be 

```
recuser = getBytesFromPrefs(service, DELIMITER + "u");
recpass = getBytesFromPrefs(service, DELIMITER + "p");
```

vonovak commented:
@pcoltau the above changes should make the transition from old version of the package to new smooth. Also, may I ask you to refactor the places in code that contain "u" and "p" strings such as 

```
recuser = getBytesFromPrefs(service, DELIMITER + "u");
recpass = getBytesFromPrefs(service, DELIMITER + "p");
```

into separate methods so that we don't have to think about where to add "u" and "p"? (it's getting confusing). Thanks!

oblador commented:
@vonovak @pcoltau I think we shouldn't break existing consumers of this library unless we have a very good reason. There's still a non-neglible size in the pre 18 segment and very many users pre 23. While I think it's OK to keep 16 support in the 0.x versions and 18+ in 1.x versions, dropping pre 23 is premature IMHO. We could encourage it in the readme, but I don't want it to be a requirement to use this library.

oblador commented:
@ggozad You're very welcome to contribute, a review of this PR would be a great start! 😄

pcoltau commented:
Thank you for all the great feedback!

I have addressed all comments and bugs regarding the current implementation, but please keep reviewing and commenting :-)

Regarding API levels less than 23 - I'm currently working on reintroducing the implementation using Facebook's Conceal for devices that use API levels less than 23, but it will likely not finish before Sunday, as the weekend is starting now in my region.

pcoltau commented:
I have updated the PR.

I did not realize the importance of backwards compatibility, so in the latest commits I have taken the consequence and made a more generic solution: 

I've added a `CipherStorage` interface and two implementations of that interface: `CipherStorageFacebookConceal` and `CipherStorageKeystoreAESCBC`. The first is the "legacy" implementation and the latter is the new using the Keystore.

The `KeychainModule` will now automatically use the appropriate `CipherStorage` implementation based on API level:

API level 16-22 will use `CipherStorageFacebookConceal`,
API level 23+ will use `CipherStorageKeystoreAESCBC`.

The `KeychainModule` will also automatically migrate values written by Facebook Conceal to the appropriate `CipherStorage` implementation.

This solution is also more future proof - whenever Android adds a newer and stronger encryption algorithm (or change the Keystore), we can just add a new implementation of `CipherStorage` and the library will automatically migrate.

Please give it a test and comment if there is anything missing.

oblador commented:
@vonovak Can you do a final review before we merge? This seems really promising to me, thanks for the great work! Care to help us maintain the android version @pcoltau?

vonovak commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -37,99 +41,108 @@ public String getName() {
+                // store the encryption result
+                prefsStorage.storeEncryptedEntry(service, encryptionResult);
+                // clean up the old cipher storage
+                oldCipherStorage.removeKey(service);
```
for `CipherStorageFacebookConceal`, this will be a no-op. But we want the conceal values be removed from sharedPreferences.

`removeKey` removes the entry from KeyStore on the new `CipherStorageKeystoreAESCBC` implementation of `CypherStorage` interface but does nothing for the `CipherStorageFacebookConceal` implementation of the `CypherStorage` interface. Then, there's the `resetGenericPasswordForOptions` method which calls `removeKey` and also calls `resetPassword` which removes the entries from sharedPreferences.

I suggest we merge the `removeKey` and `resetPassword` into one method because if I understand it correctly, when we remove a key we want to (1) remove it from the sharedPreferences and(2) from the Keystore (if using the KeyStore implementation) and we want this to happen together in one place rather than having it in multiple places.

In other words, when you call `oldCipherStorage.removeKey(service);` (where oldCipherStorage means conceal) you want this call to remove the entry from sharedPreferences. That leads me to thinking that maybe `PrefsStorage prefsStorage` should not be in `KeychainModule` but in the implementations of the `CipherStorage` interface. (Also since `PrefsStorage` has is a utility class, maybe everything in it could be declared static but that is not a big deal now). I would also allow the calls from `KeychainModule` to be more clean - not having to deal with the details of how things are stored, but having the `CipherStorage` implementations deal with it. Then you would just call something like `currentCipherStorage.encryptAndPersist(service, username, password);`

pcoltau replied:
I do not necessarily agree with the idea of moving the usage of `PrefsStorage` into each implementation of `CipherStorage`. 

The persistence of the encrypted data is independent of the actual encryption and will be the same for all implementations of `CipherStorage`. A new implementation of `CipherStorage` would then need to make the same calls to `PrefsStorage` as the other `CipherStorage` implementations.

For Facebook's Conceal, you are right, the `removeKey` is a no-op. But I assume for future implementations of `CipherStorage`, we will see it necessary to clean up keys. 

From an architectural point of view, I would prefer that we keep the separation of `CipherStorage` and `PrefsStorage`.

vonovak replied:
ok, let's keep it, but now we will need to change this to

```
CipherStorage oldCipherStorage = getCipherStorageByName(resultSet.cipherStorageName);
// decrypt using the older cipher storage
decryptionResult = oldCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
// encrypt using the current cipher storage
EncryptionResult encryptionResult = currentCipherStorage.encrypt(service, decryptionResult.username, decryptionResult.password);
// clean up the old cipher storage
oldCipherStorage.removeKey(service);
prefsStorage.resetPassword(service); //<--- this is important, to remove the old stuff from sharedPreferences
// store the encryption result
prefsStorage.storeEncryptedEntry(service, encryptionResult);
```

and ideally factor this out into a separate method, with some meaningful name.

pcoltau replied:
I did not think that it was necessary to call `prefsStorage.resetPassword()` since the call to `prefsStorage.storeEncryptedEntry()` would overwrite any existing values. Can you see a scenario were it would not overwrite?

Also, I have the call to `oldCipherStorage.removeKey()` after `prefsStorage.storeEncryptedEntry()` in case the storing fails. If it does fail, then the old key is not removed.

vonovak replied:
ah, of course, you're right

pcoltau commented on android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorage.java:
```diff
@@ -0,0 +1,41 @@
+
+    String getCipherStorageName();
+
+    int getAPILevel();
```
Sure - I'll fix.

pcoltau commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -37,99 +41,108 @@ public String getName() {
 
-    private String encryptWithEntity(String toEncypt, Entity entity, Promise promise) {
+    @ReactMethod
+    public void setGenericPasswordForOptions(String service, String username, String password, Promise promise) {
```
You are right. I wanted a null `service` to be handled internally in each CipherStorage implementation, because it might differ from implementation to implementation - but of course the prefStorage should then also handle a null `service`.

pcoltau commented on android/src/main/java/com/oblador/keychain/KeychainModule.java:
```diff
@@ -37,99 +41,108 @@ public String getName() {
+            }
+            CipherStorage currentCipherStorage = getCipherStorageForCurrentAPILevel();
+            if (currentCipherStorage == null) {
+                throw new CryptoFailedException("Unsupported Android SDK " + Build.VERSION.SDK_INT);
```
I agree.

pcoltau commented on android/src/main/java/com/oblador/keychain/exceptions/KeyStoreAccessException.java:
```diff
@@ -0,0 +1,11 @@
+package com.oblador.keychain.exceptions;
+
+public class KeyStoreAccessException extends Exception {
+    public KeyStoreAccessException(Throwable t) {
```
You are right.

pcoltau commented:
@oblador 

I would happily help maintain this library - thank you for the invite. 

@vonovak 

I have commented on all your issues and I have fixed all that I agree with :-) Let me know what you think..

oblador commented:
Absolutely beautiful 👏 Sent you an invite, lets just wait for @vonovak's blessings 😄

vonovak commented on android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageFacebookConceal.java:
```diff
@@ -0,0 +1,96 @@
+        if (!crypto.isAvailable()) {
+            throw new CryptoFailedException("Crypto is missing");
+        }
+        service = service == null ? EMPTY_STRING : service;
```
there are three places where `service` enters the module and from where it is passed onward - the methods for `get`, `set` and `reset` password. These are unlikely to change. I suggest that the `service = service == null ? EMPTY_STRING : service;` is done there so that we're really sure that we don't have to do a null check in any other methods of the module - for example in `getEntityPrefix`. It also means less duplication, since now all implementations of `CipherStorage` have to do this check.

pcoltau replied:
I agree that there is a lot of copy-pasting regarding the null-check of service.

Defaulting to the empty string is used both in the `PrefsStorage` (because that's the way it used to be) and in the `CipherStorageFacebookConceal` (for the same reason), but the empty string cannot be used in the `CipherStorageFacebookAESCBC` because it is not allowed to use empty strings as keys in the Keystore. That is why I did not keep the null-check in the `KeychainModule` as it was before.

I could add some private methods in both `CipherStorageFacebookConceal` and `CipherStorageFacebookAESCBC` that does the null-check and defaulting..?

vonovak replied:
hmm, it probably wasn't a very good idea to default to empty string in the first place :/ how about the null checks are done in the three places mentioned above, and the `CipherStorageKeystoreAESCBC` changes the `service` to some other string when it gets an empty one? Or the other way round - we if service is null we default to some nonempty string `S` and the conceal implementation changes service to an empty string if it meets `S`? Maybe better.?

pcoltau replied:
I would like to keep having the empty string in both conceal and in the PrefsStorage - otherwise we would need to migrate old values stored under the empty string to the new non-empty string - I would like to avoid doing such migration as it makes everything more complex.

I will try to implement your first proposal and see how it feels :-)

pcoltau replied:
I've committed the changes required by your first proposal - please review.

vonovak replied:
ok, let me just test it out and I think we'll be god to merge

vonovak commented on android/src/main/java/com/oblador/keychain/PrefsStorage.java:
```diff
@@ -0,0 +1,110 @@
+        return null;
+    }
+
+    public void resetPassword(String service) {
```
can you please rename this to `removePreferencesEntry` or something like that?

pcoltau replied:
Good idea - I'll do that.

oblador commented:
Good to merge @vonovak?

vonovak commented:
I tested it out and works for me! great work @pcoltau

oblador commented:
I'll let you have the honor of merging @pcoltau

MacKentoch commented:
Awesome! thank you for this PR @pcoltau 👍

chris-yoon90 commented:
Great stuff! I've been waiting for this pull request to be merged. Is there any plans to release the next version with this PR anytime soon? Thanks for the hard work!

oblador commented:
Released an RC with this functionality, would be grateful for anybody testing it thoroughly before we can release as stable. Test it with `yarn add react-native-keychain@2.0.0-rc`

vonovak commented:
oh jeez, now I noticed the commits weren't squashed :/ sorry for polluting the history

EJohnF commented:
Actually, we meet a problems with this version (react-native-keychain@2.0.0-rc)

"E/RNKeychainManager(19040): Could not decrypt bytes"

on devices with Android OS versions: 7.0, 8.0, 8.1

vonovak commented:
@EJohnF hi! thanks for reporting, can you please create a minimal reproduction and file an issue for it? Thanks!

EJohnF commented:
Sorry, personally I also can't reproduce, just get back from a customer who records a logcat for me with an error message I put above.

So, I don't sure which part and why exactly fails, we have just two simple usage of lib:

return Keychain.setGenericPassword(email, result.token);
--
const credentials = await Keychain.getGenericPassword();

vonovak commented:
@asinel if you have a problem and you're able to clearly reproduce it then please create an issue for it! Without reproductions there isn't much that can be done.
